integration_tests/contracts/contracts/TestERC20A.sol

```diff
@@ -2,7 +2,14 @@ pragma solidity ^0.6.6;
 import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 
 contract TestERC20A is ERC20 {
+	event __CronosNativeTransfer(address recipient, uint256 amount);
+	event __CronosEthereumTransfer(address recipient, uint256 amount, uint256 bridge_fee);
+
 	constructor() public ERC20("Bitcoin MAX", "MAX") {
 		_mint(msg.sender, 100000000000000000000000000);
 	}
+
+	function test_native_transfer(uint amount) public {
+		emit __CronosNativeTransfer(msg.sender, amount);
```

Comment:
Don't u have to burn the amount?

imo `__CronosNativeTransfer` is not clear for the operation. A transfer imply an origin and a destination. 

Here it seems that you are either converting or minting?

It seems more that you are con

Reply:
This is solely for testing the native call, the semantic of the call is to transfer mapped native token from contract address to recipient.
But because we can't setup token mapping in e2e testing right now, so the test itself is not complete yet (there's a TODO in that).

Reply:
so maybe `__CronosCosmosConvert` or `__CronosCosmosMint`?

Reply:
> so maybe `__CronosCosmosConvert` or `__CronosCosmosMint`?

“Native” stands for native token, “Transfer” is also self explanatory I believe?

Reply:
`is also self explanatory I believe?`

I dont think so..
"Native" has barely any meaning without context. And using "Transfer" without specifying a sender and destination sounds weird. 

If we can't find a proper short term to describe the operation, I would prefer to be very explicit and call the event 
"__CronosSendFromContractToAccount(receiver, amount)" instead...

Reply:
> `is also self explanatory I believe?`
> 
> I dont think so..
> "Native" has barely any meaning without context. And using "Transfer" without specifying a sender and destination sounds weird.
> 
> If we can't find a proper short term to describe the operation, I would prefer to be very explicit and call the event
> "__CronosSendFromContractToAccount(receiver, amount)" instead...
> 
> Mind that most smart contract developer won't have deep knowledge about the internal implementation, or cosmos implementation

I think the full version should be `__CronosSendNativeTokensFromContractToAccount` 😄

Reply:
I agree with @thomas-nguy ... generally the best is we have a concise term, sometimes involve some domain knowledge, for the action; if not better be explicit than ambiguous.

Reply:
How about `__CronosNativeTokenTransfer`, if we look at the signature of the event, it's clear the first argument is the recipient:
```
event __CronosNativeTokenTransfer(address recipient, uint256 amount);
```

Also any suggestion to the name of `__CronosEthereumTransfer`?

Reply:
In fact, the name CronosSendNativeTokensFromContractToAccount may still be confusing. Perhaps I overlook sth. But you don't have native tokens in a smart contract, so how can you send native tokens from contract to account? So at the end the name is still not self-explanatory?

But I got a smell that maybe it's over-correct for now 😂 because @yihuang you mentioned this test case is incomplete, so the true meaning may change again when work on it later.

In the code below it seems that it is trying to mint a new native token and credit to the user account?

Reply:
when I read this line 
```__CronosNativeTokenTransfer(msg.sender, amount);```

I interpret it as "I am sending CRO amount from my account to the contract"

I think it is worth to spend some time so that we can have a consensus on the cronos events because they will be a standard for our platform in the future

Reply:
actually the correct term would be 
``` __CronosSendCoinsFromContractToAccount ``` 

Because bank module use the term coin, not native token,  but since we don't specify the coin denom 

A more accurate term would be 

``` __CronosSendCosmosContractCoinFromContractToAccount ```  but well, its quite long

Reply:
> In fact, the name CronosSendNativeTokensFromContractToAccount may still be confusing. Perhaps I overlook sth. But you don't have native tokens in a smart contract, so how can you send native tokens from contract to account? So at the end the name is still not self-explanatory?
> 
> But I got a smell that maybe it's over-correct for now 😂 because @yihuang you mentioned this test case is incomplete, so the true meaning may change again when work on it later.
> 
> In the code below it seems that it is trying to mint a new native token and credit to the user account?

The meaning itself is clear here, it's just transformed to the call: `bankKeeper.SendCoins(contract, recipient, sdk.NewCoin(denom, amount))`.

Reply:
Actually maybe we should change the event to be 
```
__CronosSendCoinFromContractToAccount(recipient, amount, denom)

```

That way, the trigger can be generic. You can send "Any coin" from the contract address to the recipient account, not only the gravity linked coin

Reply:
I guess my habit of using native tokens coming from [cardano](https://developers.cardano.org/docs/native-tokens/).

Reply:
> and the contract address should have the native tokens already.
Ohhhh, ok so I think this is the context I have overlooked before. Does it mean every contract will keep its own Cosmos native token?

Reply:
> > and the contract address should have the native tokens already.
> 
> Ohhhh, ok so I think this is the context I have overlooked. Does it mean every contract will keep its own Cosmos native token?

Yeah, when one converts native tokens to erc20 tokens, those tokens are locked under the contract address.

Reply:
> I think it is worth to spend some time so that we can have a consensus on the cronos events because they will be a standard for our platform in the future

I agree, maybe we should have a separate issue for this discussion. That would better record our decision making process.

> Because bank module use the term coin, not native token, but since we don't specify the coin denom

Yes, I think we also need to better define these terms as well. I do use `native` all around to represent token that exists at the Cosmos level. i.e. Coin. But maybe this is not the convention in the Cosmos community. Have to be careful about this.

Reply:
> I agree, maybe we should have a separate issue for this discussion. That would better record our decision making process.

 https://github.com/crypto-org-chain/cronos/issues/60
 Opened an issue for this.

Reply:
Thanks @yihuang 

Shall we discuss event naming pattern in the same issue or a separate one?

Reply:
`__CronosXTransferFromContract(address recipient, uint256 amount)`

Assume `X` here stands for the cosmos native token denom that is associated with the contract, I guess this pattern is clear enough?

Reply:
When I think about it again, can we say "From" is implied because the action is "Send" and we are executing a contract? It is the "Subject" (what token) and "To" that's unclear.

So
```
___CronosSendXToAccount // Account or Native Account? need to address in the terminology issue
___CronosSendXToEthereum // Send X (From the contract account) To Ethereum
___CronosSendXToIBCChannel
// Similar to Gravity Bridge `SendToCosmos` but be more explicit about the token type
```

p.s. Having "From" is no harm, my motive is if we think event name is too long then we can think about to trim it or not. I am okay to include it and be more explicit. (Is there any restricting on event name length?)

Reply:
> When I think about it again, can we say "From" is implied because the action is "Send" and we are executing a contract? It is the "Subject" (what token) and "To" that's unclear.
> 
> So
> 
> ```
> ___CronosSendXToAccount // Account or Native Account? need to address in the terminology issue
> ___CronosSendXToEthereum // Send X (From the contract account) To Ethereum
> ___CronosSendXToIBCChannel
> // Similar to Gravity Bridge `SendToCosmos` but be more explicit about the token type
> ```
> 
> p.s. Having "From" is no harm, my motive is if we think event name is too long then we can think about to trim it or not. I am okay to include it and be more explicit. (Is there any restricting on event name length?)

Looks good to me.
And let's use `NativeToken` for `X` for now?

Reply:
Maybe 'X' can be omitted now?

__CronosSendToAccount
__CronosSendToEthereum
__CronosSendToIBCChannel

Even if we add `NativeToken` or `Coins`, it is not clear which coin is it. We need a proper term to designate the "Coin associated with the contract"

Reply:
> Maybe 'X' can be omitted now?
> 
> __CronosSendToAccount
> __CronosSendToEthereum
> __CronosSendToIBCChannel
> 
> Even if we add `NativeToken` or `Coins`, it is not clear which coin is it. We need a proper term to designate the "Coin associated with the contract"

Looks good too, I suppose we'll have detailed documentation about them as part of the CRC20 standard.

Reply:
https://github.com/crypto-org-chain/cronos/issues/62
opened issue and documented decision so far in it.